promtimer/annotations.py

```diff
@@ -0,0 +1,181 @@
+    'XDCR_replication_create_successful': 'success',
+    'XDCR_replication_created': 'success',
+    'XDCR_replication_remove_failed': 'failure',
+    'XDCR_replication_remove_successful': 'success',
+}
+
+def retry_get(url, retries):
+    req = urllib.request.Request(url=url, data=None)
+    success = False
+    get = None
+    while (not success) and (retries > 0):
+        try:
+            get = urllib.request.urlopen(req).read()
+            success = True
+        except:
+            print('Attempting connection to Grafana, retrying...',retries,'retries left')
+            retries -= 1
+            time.sleep(0.5)
+    return get
+
+def create_annotations():
+    annotations_json = retry_get(ANNOTS_URL, 5)
+    if annotations_json is not None:
+        print('Successfully connected to Grafana')
+        annotations_json = json.loads(annotations_json)
+        if len(annotations_json) > 0:
+            print('Found existing annotations, skipping annotation adding')
+        else:
+            if not path.isfile('events.log'):
+                print('No events.log, skipping annotation adding')
+            else:
+                print('Adding annotations from events.log')
+                ongoing_events = {}
+                file = open('events.log', 'r')
+                for line in file:
+                    event = json.loads(line)
+                    event_timestamp = event['timestamp']
+                    event_type = event['event_type']
+                    unix_time_ms = int(dateparser.parse(event_timestamp).timestamp()*1000)
+                    try:
+                        tag = ANNOTS_EVENT_TAGS[event_type]
+                        if event_type in ANNOTS_EVENTS_START:
+                            ongoing_events[ANNOTS_EVENTS_START[event_type]] = unix_time_ms
+                            continue
+                        elif event_type in ANNOTS_EVENTS_END and ANNOTS_EVENTS_END[event_type] in ongoing_events:
+                            data = {
+                                'time': ongoing_events[ANNOTS_EVENTS_END[event_type]],
+                                'timeEnd': unix_time_ms,
+                                'text': ANNOTS_EVENTS_END[event_type],
+                                'tags': [
+                                    tag,
+                                ],
+                            }
+                            ongoing_events.pop(ANNOTS_EVENTS_END[event_type])
+                        else:
+                            data = {
+                                'time': unix_time_ms,
+                                'text': event_type,
+                                'tags': [
+                                    tag,
+                                ],
+                            }
+                        payload = json.dumps(data).encode('utf-8')
+                        req = urllib.request.Request(url=ANNOTS_URL, data=payload, headers=ANNOTS_HEADERS)
```

Comment:
Create a separate function that does the post against the annotations URL. You've a bunch of copy and pasted code.